kong/clustering/version_negotiation/init.lua

```diff
@@ -0,0 +1,340 @@
+    services_rejected = services_rejected,
+  }
+end
+
+
+local function register_client(conf, client_node, services_accepted)
+  local services_map = {}
+  for _, service in ipairs(services_accepted) do
+    services_map[service.name] = service.version
+  end
+
+  local ok, err = kong.db.clustering_data_planes:upsert({ id = client_node.id, }, {
+    last_seen = ngx.time(),
+    config_hash = "00000000000000000000000000000000",
+    hostname = client_node.hostname,
+    ip = ngx.var.remote_addr,
+    version = client_node.version,
+    sync_status = client_node.sync_status,
+    services_accepted = services_map,
+  }, { ttl = conf.cluster_data_plane_purge_delay })
+
+  if not ok then
+    ngx_log(ngx_ERR, _log_prefix, "unable to update clustering data plane status: ", err)
+    return nil, err
+  end
+
+  return true
+end
+
+--- Handles a version negotiation request (CP side).
+--- Performs mTLS verification (as configured),
+--- validates request and Kong version compatibility,
+---
+function _M.serve_version_handshake(conf, cert_digest)
+  if KONG_VERSION == nil then
+    KONG_VERSION = kong.version
+  end
+
+  local ok, err = clustering_utils.validate_connection_certs(conf, cert_digest)
+  if not ok then
+    ngx_log(ngx_ERR, _log_prefix, err)
+    return ngx.exit(ngx.HTTP_CLOSE)
+  end
+
+  local body_in = cjson.decode(get_body())
+  if not body_in then
+    return response_err("Not valid JSON data")
+  end
+
+  ok, err = verify_request(body_in)
+  if not ok then
+    return response_err(err)
+  end
+
+  ok, err, body_in.node.sync_status = check_node_compatibility(body_in.node)
+  if not ok then
+    return response_err(err)
+  end
+
+  local body_out = do_negotiation(body_in)
+
+  ok, err = register_client(conf, body_in.node, body_out.services_accepted)
+  if not ok then
+    return response_err(err)
+  end
+
+  return response(200, body_out)
+end
+
+--- Performs version negotiation request (DP side).
+--- Stores the responses to be queried via get_negotiated_service(name)
+--- Returns the DP response as a Lua table.
+function _M.request_version_handshake(conf, cert, cert_key)
+  local body = cjson.encode{
+    node = {
+      id = kong.node.get_id(),
+      type = "KONG",
+      version = kong.version,
+      hostname = kong.node.get_hostname(),
+    },
+    services_requested = require "kong.clustering.version_negotiation.services_requested",
+  }
+
+  local params = {
+    scheme = "https",
+    method = "POST",
+    headers = {
+      ["Content-Type"] = "application/json",
+    },
+    body = body,
+
+    ssl_verify = false,
+    ssl_client_cert = cert,
+    ssl_client_priv_key = cert_key,
+  }
+  if conf.cluster_mtls == "shared" then
+    params.ssl_server_name = "kong_clustering"
+  else
+    -- server_name will be set to the host if it is not explicitly defined here
+    if conf.cluster_server_name ~= "" then
+      params.ssl_server_name = conf.cluster_server_name
+    end
+  end
+
+  local c = http.new()
```

Comment:
Maybe we can use connection passed from an argument, so that later communication can be done through the same connection, like `Keep-Alive`?

Reply:
Not sure if this is proper but it's just an idea: https://github.com/openresty/lua-resty-websocket/issues/73